tests/test_settings.py

```diff
@@ -0,0 +1,38 @@
+import logging
+
+import teachers_gitlab.main as tg
+import gitlab
+
+def test_project_settings(mock_gitlab):
+    entries = [
+        {'login': 'alpha'},
+    ]
+
+    mock_gitlab.register_project(42, 'student/alpha', mr_default_target_self='self')
+
+    tg.action_project_settings(
+        mock_gitlab.get_python_gitlab(),
+        logging.getLogger("settings"),
+        tg.ActionEntries(entries),
+        True,
```

Comment:
Here should be `False`: otherwise the test runs in _dry run_ mode, i.e. it does not perform any modifications on GitLab. I also believe that you should call `mock_gitlab.report_unknown()` somewhere to ensure that you do not miss any API calls.

Reply:
Thanks for the advice, however, I tried doing it the way you said and an issue emerged. Even after the `mock_gitlab.report_unknown()` I was not able to mock call `PUT http://localhost/api/v4/projects/42`, which isn't even in the registered responses right?

Reply:
Right, the first time we need the `PUT` method ;-).

I think the following should help as it will add `on_api_put` that behaves similarly to `on_api_post` and should allow you to mock the PUT call modifying the project settings.

```diff
diff --git a/tests/mocks_gitlab_api.py b/tests/mocks_gitlab_api.py
index 2d33c00..e59ef97 100644
--- a/tests/mocks_gitlab_api.py
+++ b/tests/mocks_gitlab_api.py
@@ -22,7 +22,7 @@ class MockedGitLabApi:
             self.unknown_urls.append(log_line)
             return (404, {}, json.dumps({"error": "not implemented"}))
 
-        methods = [responses.GET, responses.POST, responses.DELETE]
+        methods = [responses.GET, responses.POST, responses.DELETE, responses.PUT]
         for m in methods:
             self.responses.add_callback(
                 m,
@@ -97,3 +97,16 @@ class MockedGitLabApi:
             *args,
             **kwargs,
         )
+
+    def on_api_put(self, url, request_json, response_json, *args, **kwargs):
+        kwargs['body'] = json.dumps(response_json)
+        kwargs['match'] = [
+            responses.matchers.json_params_matcher(request_json)
+        ]
+        kwargs['content_type'] = 'application/json'
+
+        return self.responses.put(
+            self.make_api_url_(url),
+            *args,
+            **kwargs,
+        )
```

```diff
diff --git a/tests/test_settings.py b/tests/test_settings.py
index 4f00904..1e416e4 100644
--- a/tests/test_settings.py
+++ b/tests/test_settings.py
@@ -26,12 +26,15 @@ def test_project_settings_with_none(mock_gitlab):
     ]
 
     mock_gitlab.register_project(38, 'student/beta', mr_default_target_self='self')
+    mock_gitlab.on_api_put('projects/42', ...)
+
+    mock_gitlab.report_unknown()
 
     tg.action_project_settings(
         mock_gitlab.get_python_gitlab(),
         logging.getLogger("settings"),
         tg.ActionEntries(entries),
-        True,
+        False,
         'student/{login}',
         None,
         'The best project'
```

Instead of the three dots in `mock_gitlab.on_api_put('projects/42', ...)` we need the actual JSON that is sent/returned when updating project settings (at least part of it).